extensions/chathistory.md

```diff
@@ -0,0 +1,106 @@
+
+#### Context
+
+The following keywords are used to describe how the server should return messages relative to the `timestamp(s)` or `draft/msgid(s)` given.
+
+`BEFORE` get up to `limit` number of messages before the given `timestamp` or `draft/msgid`. The `limit` MUST BE positive.
+
+`AFTER` get up to `limit` number of messages after the given `timestamp` or `draft/msgid`. The `limit` MUST BE positive.
+
+`LATEST` get the most recent (up to `limit`) messages that have been sent since the given `timestamp` or `draft/msgid`. The `limit` MUST BE positive.
+
+`AROUND` get a number of messages before and after the `timestamp` or `draft/msgid`. The `limit` MUST BE positive.
+
+`BETWEEN` get up to `limit` number of messages between the given `timestamps` or `draft/msgids`. The `limit` MAY positive or negative. If the `limit` is positive, messages will be returned in ascending order starting at the oldest message. If the `limit` is negative, messages will be returned in decending order starting at the newest message. 
+
+#### Format
+`chathistory` uses the following generic format:
+
+    @draft/label=ID CHATHISTORY <target> <context> <start> <end> [<direction>] [<limit>]
+
+The `chathistory` content can requested using timestamps:
+
+    @draft/label=ID CHATHISTORY <target> <context> timestamp=YYYY-MM-DDThh:mm:ss.sssZ [<direction>] [<limit>]
+
+Alternatively, content can be requested using a `draft/msgid`:
+
+    @draft/label=ID CHATHISTORY <target> <context> draft/msgid=<message_id> timestamp=<timestamp>
+
+For `BEFORE`, `AFTER`, `LATEST`, and `AROUND`, a single `timestamp` or `draft/msgid` is required.  For `BETWEEN`, both a `start` and `end` `timestamp` and/or `draft/msgid` is required. The  `limit` is optional for all contexts.
+
+#### Errors and Warnings
+If the server receives a `CHATHISTORY` command with missing parameters, the `NEED_MORE_PARAMS` error code SHOULD be returned.
+
+If the number of lines between the `start` and `end` parameters exceeds the `max_messages`, warn code `MAX_MESSAGES_EXCEEDED` SHOULD be returned. The command SHOULD continue to be processed as described above.
+
+If the target has no `chathistory` content to return or the user does not have permission to view the requested content, `NO_SUCH_NICK` or `NO_SUCH_CHANNEL` SHOULD be returned accordingly.
+
+### Examples
+The examples below are written with `draft/msgid` and `draft/label` tags included. These tags are recommended.
+
+    @draft/label=ID :irc.host BATCH +ID chathistory target
+
+    @batch=ID;draft/msgid=ID;time=YYYY-MM-DDThh:mm:ss.sssZ :nick!ident@host PRIVMSG target :message
+
+    @batch=ID;draft/msgid=ID;time=YYYY-MM-DDThh:mm:ss.sssZ :nick!ident@host NOTICE target :message
+
+    @batch=ID;draft/msgid=ID;time=YYYY-MM-DDThh:mm:ss.sssZ :nick!ident@host PRIVMSG target :ACTION message
+
+    :irc.host BATCH -ID
+
+    @draft/label=ID :nick!ident@host ERR CHATHISTORY ERR_CODE
```

Comment:
Clients should be able to associate the `WARN_CODE` / `ERR_CODE` with the specific chathistory return that's failing/warning. Feels like they should instead be inside the batch, rather than separately outside it. There's no compatibility issues with including 'em inside the batch because the only clients that'll receive these messages are ones that already support `CHATHISTORY`.

Reply:
We also need to define the top level err/warn more fully. Do they allow just free form text as a generic alert or do they always need a command and error code?

Reply:
#357 submitted, specifies how I think these sort of response messages should work.

Reply:
I agree the error needs to be inside of the batch, that way you can at least "continue".